C API `wasmtime_val_t` layout mismatch on 32-bit platforms (i686) due to union alignment

### Issue Description

On 32-bit platforms (e.g., `i686-linux-android`, `armv7-linux-androideabi`), the C API header `wasmtime/val.h` contains `static_assert` checks that require `wasmtime_valunion_t` and `wasmtime_val_raw_t` to be 8-byte aligned:

```c
static_assert(__alignof(wasmtime_valunion_t) == 8, "should be 8-byte aligned");
static_assert(__alignof(wasmtime_val_raw_t) == 8, "should be 8-byte aligned");
```

However, on 32-bit platforms, `int64_t` and `double` only have 4-byte alignment in the C ABI, causing these assertions to fail at compile time.

Additionally, even if the assertions are bypassed, there is a **layout mismatch** between Rust and C on 32-bit platforms:

| | Rust `#[repr(C)]` | C (original `val.h`) |
|---|---|---|
| union alignment | 8 (Rust's `i64`/`u64` are always 8-byte aligned) | 4 (C ABI `int64_t` on 32-bit) |
| `wasmtime_val_t.of` offset | 8 | 4 |
| `wasmtime_val_t` size | 32 | 20 |

This causes the C side to read garbage data for the `kind` field (e.g., `unknown wasmtime_valkind_t: 16`), leading to a `SIGABRT` crash at runtime when calling `wasmtime_func_call`.

### Reproduction

1. Build `libwasmtime.a` for `i686-linux-android` (32-bit x86 Android)
2. Include `wasmtime/val.h` from C++ code compiled with the Android NDK (x86 target)
3. Compile fails with `static_assert` errors
4. If assertions are removed, runtime crash occurs: `Abort message: 'unknown wasmtime_valkind_t: <garbage>'`

### Crash Log (x86 Android Emulator)

```
Abort message: 'unknown wasmtime_valkind_t: 16'
signal 6 (SIGABRT), code -1 (SI_QUEUE)
backtrace:
  #14 pc ... libwasmline.so (wasmtime_func_call+192)
  #15 pc ... libwasmline.so (wasmline::Session::invokeInbound+465)
  ...
```

### Proposed Fix

Force 8-byte alignment on both C and Rust sides to ensure consistent layout across all platforms:

**`crates/c-api/include/wasmtime/val.h`:**
```diff
-typedef union wasmtime_valunion {
+typedef union alignas(8) wasmtime_valunion {

-typedef union wasmtime_val_raw {
+typedef union alignas(8) wasmtime_val_raw {

-typedef struct wasmtime_val {
+typedef struct alignas(8) wasmtime_val {
```

**`crates/c-api/src/val.rs`:**
```diff
-#[repr(C)]
+#[repr(C, align(8))]
 pub struct wasmtime_val_t {

-#[repr(C)]
+#[repr(C, align(8))]
 pub union wasmtime_val_union {
```

And update the compile-time assertion:
```diff
 const _: () = {
-    assert!(std::mem::size_of::<wasmtime_val_union>() <= 24);
-    assert!(std::mem::align_of::<wasmtime_val_union>() == std::mem::align_of::<u64>());
+    assert!(std::mem::size_of::<wasmtime_val_union>() <= 32);
+    assert!(std::mem::align_of::<wasmtime_val_union>() == 8);
 };
```

### Impact

| Platform | Before Fix | After Fix |
|----------|-----------|-----------|
| x86_64 / aarch64 (64-bit) | align=8, works ✅ | align=8, no change ✅ |
| x86 / armv7a (32-bit) | align=4, compile fails / runtime crash ❌ | align=8, works ✅ |

- **64-bit platforms**: zero impact (`alignas(8)` is redundant, natural alignment is already 8)
- **32-bit platforms**: fixes compile-time assertion failure and runtime layout mismatch
- **Memory overhead on 32-bit**: ~12 bytes per `wasmtime_val_t` (negligible for typical usage)
- **API compatibility**: no changes to type names or function signatures

### Environment

- Forked Wasmtime version: 45.0.6 (custom features pulley-min [x86/armv7a] )
- Affected targets: `i686-linux-android`, `armv7-linux-androideabi` (and potentially other 32-bit targets)
- Discovered while building wasmline (a Wasm plugin framework) with Pulley interpreter on 32-bit Android

### Fix
- https://github.com/crowforkotlin/wasmtime/commit/e24159c780779c7f36554c7578b94bd5b37708a0#diff-ed790182a50ae9da596d84eb7f31ac26eb4af8c326a9d19a637ab903c7a2091b

### Result After Fix

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/e9c742d0-a673-4eae-b8eb-16e4ac49973e" />

> Can you share some more information about how to reproduce this? [Local testing](https://godbolt.org/z/E8s9dfK95) shows that the align of u64/f64 are both 8 on arm platforms. I know it can be 4 sometimes, but I'm not sure how you're seeing 4 on arm platforms.
> 
> One possible difference is that Rust uses `u32`/`u64` for the f32/f64 wasm fields, but the C side uses `float32_t` and `float64_t` which means that if the align of `u64` is not equal to `f64` this'll cause problems. Before I fix that though I'd want to better understand how to reproduce and where the align 4 is coming from.

```
In file included from /home/crowf/Downloads/error-result/pulley-runner.cpp:12:
In file included from /home/crowf/Downloads/error-result/wasmtime-dev-x86-android-c-api/include/wasmtime.h:199:
In file included from /home/crowf/Downloads/error-result/wasmtime-dev-x86-android-c-api/include/wasmtime/anyref.h:12:
/home/crowf/Downloads/error-result/wasmtime-dev-x86-android-c-api/include/wasmtime/val.h:353:17: error: static assertion failed due to requirement '__alignof(wasmtime_valunion) == 8': should be 8-byte aligned
  353 |   static_assert(__alignof(wasmtime_valunion_t) == 8,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/crowf/Downloads/error-result/wasmtime-dev-x86-android-c-api/include/wasmtime/val.h:353:48: note: expression evaluates to '4 == 8'
  353 |   static_assert(__alignof(wasmtime_valunion_t) == 8,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/home/crowf/Downloads/error-result/wasmtime-dev-x86-android-c-api/include/wasmtime/val.h:356:17: error: static assertion failed due to requirement '__alignof(wasmtime_val_raw) == 8': should be 8-byte aligned
  356 |   static_assert(__alignof(wasmtime_val_raw_t) == 8, "should be 8-byte aligned");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/crowf/Downloads/error-result/wasmtime-dev-x86-android-c-api/include/wasmtime/val.h:356:47: note: expression evaluates to '4 == 8'
  356 |   static_assert(__alignof(wasmtime_val_raw_t) == 8, "should be 8-byte aligned");
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```

> To clarify, `align(8)` shouldn't be necessary and it's not desired. The previous assertions were just buggy and needed fixing, the code didn't need other changes.
> 
> For reproduction, yeah I understand that minimization isn't always easy. That being said there's a chasm of the understanding of wasmtime maintainers and what you're working on, and somehow that chasm needs to be bridged. On one hand you can cross the chasm by creating a minimal reproduction, but on the other hand we could put in a lot of time to learn what you're doing and cross the chasm as well. Who's got the time depends on a whole bunch of factors, but I want to point out primarily that it's quite a lot of work to ask and/or expect of wasmtime maintainers to fully internalize such a large project (and I subjectively thought it had quite a lot of AI noise too)

That makes sense, thanks for the explanation.

To give an update, I’ve ported the fixed code back into my local codebase. While the Rust side now compiles successfully, I hit a static assertion failure during the subsequent NDK build stage. compiler complains that `wasmtime_valunion_t` and `wasmtime_val_raw_t` are evaluated to 4-byte alignment instead of the expected 8 bytes:

```text
error: static assertion failed due to requirement '__alignof(wasmtime_valunion) == 8': should be 8-byte aligned
note: expression evaluates to '4 == 8'

```

To investigate whether this is an inherent alignment difference on 32-bit x86, I wrote a minimal C program and tested it on an `i686-linux-android` target. The output confirmed that `uint64_t` and `double` (and consequently the union) are indeed 4-byte aligned on this platform.

Here is the test code and its execution output:

```c
// test_align.c — compiler target: i686-linux-android
#include <stdalign.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>

typedef union {
  int32_t i32;
  int64_t i64;
  float f32;
  double f64;
  void *ptr;
} test_union_t;

int main() {
  printf("sizeof(uint64_t)  = %zu\n", sizeof(uint64_t));
  printf("alignof(uint64_t) = %zu\n", alignof(uint64_t));
  printf("sizeof(double)    = %zu\n", sizeof(double));
  printf("alignof(double)   = %zu\n", alignof(double));
  printf("sizeof(void*)     = %zu\n", sizeof(void *));
  printf("alignof(void*)    = %zu\n", alignof(void *));
  printf("sizeof(test_union_t)  = %zu\n", sizeof(test_union_t));
  printf("alignof(test_union_t) = %zu\n", alignof(test_union_t));
  return 0;
}
```

**Output on device via adb:**

```text
sizeof(uint64_t)  = 8
alignof(uint64_t) = 4
sizeof(double)    = 8
alignof(double)   = 4
sizeof(void*)     = 4
alignof(void*)    = 4
sizeof(test_union_t)  = 8
alignof(test_union_t) = 4
```
<img width="1040" height="790" alt="Image" src="https://github.com/user-attachments/assets/b9177ec5-89a4-483a-b099-3e9d1b5f3c92" />

```shell
$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/bin/clang \
    --target=i686-linux-android23 \
    test_align.c -o test_align
adb push test_align /data/local/tmp/
adb shell chmod +x /data/local/tmp/test_align
adb shell /data/local/tmp/test_align
```

```text
$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/bin/clang \
    --target=armv7-linux-androideabi23 \
    -c test_align_v2.c -o test_align_v2.o
test_align_v2.c:12:16: error: static assertion failed due to requirement '__alignof(test_union_t) == 4': Error: test_union_t is NOT 8-byte aligned!
   12 | _Static_assert(__alignof(test_union_t) == 4,
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_align_v2.c:12:40: note: expression evaluates to '8 == 4'
   12 | _Static_assert(__alignof(test_union_t) == 4,
      |                ~~~~~~~~~~~~~~~~~~~~~~~~^~~~
1 error generated.
```
```c
// test_align_v2.c
#include <stddef.h>
#include <stdint.h>

typedef union {
  int32_t i32;
  int64_t i64;
  float f32;
  double f64;
  void *ptr;
} test_union_t;

_Static_assert(__alignof(test_union_t) == 4, "error: test_union_t is NOT 8-byte aligned!");

int main() { return 0; }
```

Given this, on 32-bit x86 Android, the natural alignment for these 64-bit types is 4. This seems to be why the `static_assert(..., 8)` fails during the NDK build unless we explicitly enforce alignment attributes in the headers. How should we best reconcile this ABI discrepancy between Rust and C on 32-bit platforms?

--- a/crates/c-api/include/wasmtime/val.h
+++ b/crates/c-api/include/wasmtime/val.h
@@ -343,6 +343,20 @@ typedef union wasmtime_val_raw {
   void *funcref;
 } wasmtime_val_raw_t;
 
+/**
+ * Portable ABI alignment operator.
+ *
+ * Uses `alignof` (C11/C++11) which returns the ABI alignment, unlike
+ * the GCC extension `__alignof` which returns the preferred alignment.
+ * Falls back to `__alignof` on MSVC which does not support `alignof`
+ * in C mode. Undef'd after use.
+ */
+#if defined(__cplusplus) || !defined(_MSC_VER)
+#define WASMTIME_ALIGNOF(t) alignof(t)
+#else
+#define WASMTIME_ALIGNOF(t) __alignof(t)
+#endif
+
 // Assert that the shape of this type is as expected since it needs to match
 // Rust.
 static inline void __wasmtime_val_assertions() {
@@ -350,12 +364,16 @@ static inline void __wasmtime_val_assertions() {
                     sizeof(wasmtime_valunion_t) <= 24,
                 "should be 16 bytes plus a pointer large (plus alignment on "
                 "some platforms)");
-  static_assert(__alignof(wasmtime_valunion_t) == __alignof(uint64_t),
+  static_assert(WASMTIME_ALIGNOF(wasmtime_valunion_t) ==
+                    WASMTIME_ALIGNOF(uint64_t),
                 "should be aligned to u64");
   static_assert(sizeof(wasmtime_val_raw_t) == 16, "should be 16 bytes large");
-  static_assert(__alignof(wasmtime_val_raw_t) == __alignof(uint64_t),
+  static_assert(WASMTIME_ALIGNOF(wasmtime_val_raw_t) ==
+                    WASMTIME_ALIGNOF(uint64_t),
                 "should be aligned to u64");
 }
+
+#undef WASMTIME_ALIGNOF
 
 /**
  * \typedef wasmtime_val_t

